### DISCUSSION TITLE
LNA design example

### DISCUSSION THREAD

**Discussion by stan3c** [scikit-rf/scikit-rf#1038] [Q&A] [answered]
There is a very nice example of an LNA calculation in skrf repository:
[LNA Example.ipynb](https://github.com/rdireen/scikit-rf/blob/50f0e13646472848f92427df64ab764d3de3b1fb/doc/source/tutorials/LNA%20Example.ipynb#L23)
At first I was unable to get it to work, so I wrote a script based on the tutorial from the above webpage and I placed my script here:
https://github.com/stan3c/skrf-LNA

I find the original Jupiter Notebook LNA tutorial page very interesting example and would want to turn it into a script that fully automates the design of a simple LNA.
My script already works but it is still messy and there are a few points that are not clear to me.
Could you please help me to get the script to work and explain a few things to me?

So, I think I understand the code up to the line nr 146 "gamma_s = 0.0" . Then things become difficult.
The function calc_matching_network_vals(z1, z2) calculates L-matching network (I think). Does skrf provide a better way of doing it? What about other types of matching networks, L-type  with first element in series, pi-matching networks?
gamma_l is calculated to be what? Optimum, max gain, max power, .... ?
I had to modify the function and define "z1_par = 1e-6" because the original "z1_par = 0.0" was rising an error. I do not understand what was the original author intention in that line.
How to write a function that calculates gamma_l for a gain of my choice, not max gain, not just conjugate impedance, but gamma_l for gain of main choice?
Do I understand correctly that c_par, l_ser are parallel and series elements of an L-matching network?

gamma_l is calculated for the case when there is no input matching network. 
How to calculate gamma_l when there is an input matching network?

Later it is written on the  Jupiter notebook page "The capacitance is kind of low but the inductance seems reasonable. Let's test it out:". I do not understand how the output network is built. In my script it is Fig. 5.

```
fig, ax = plt.subplots()
plt.title("Fig. 5 Output matching")
output_network = tem.shunt_capacitor(c_par) ** tem.inductor(l_ser)
amplifier = bjt ** output_network
amplifier.plot_s_smith()
```
In the above code, how to built an amplifier with L-input matching, and pi-output matching network?

Does the Fig. 5 make sense at all? s11 seems to be way off.

How to fix the legend in Fig.2, Fig.3, Fig.4 so they show the right thing?

Fig.6 -How to print gain as a number at any given frequency?

I hope, I am not asking too many questions....

**Comment by stan3c** [comment]
Another question.
How do I plot constant gain circles?
At the moment I use the code below, but it plots a line, not a circle. And I get a warning that there is no gain.  There is no warning if I put as gain a number smaller than 0.5.
```
tmp/ipykernel_736351/3598348151.py:1: RuntimeWarning: The specified gain is greater than the maximum gain achievable by the matching network. Specify a smaller gain.
  sgc = bjt.gain_circle(target_port=1, gain=12.0
```


```
fig, ax = plt.subplots()  
plt.title("Fig. 7 Constant gain circles (source)")
sgc = bjt.gain_circle(target_port=0, gain=12.0)
skrf.plotting.plot_smith(s=sgc, smith_r=1, marker='o', chart_type='z', draw_labels=True)
plt.show()
```

**Comment by mhuser** [comment, marked as answer]
> In the above code, how to built an amplifier with L-input matching, and pi-output matching network?

Hi @stan3c , I am not familiar with LNA design, but I think you are trying to build a network like this one:

```
0-+-L1--BJT-+-L2-+--1
  |         |    |
  C1        C2   C3
  |         |    |
 GND       GND  GND
```

```Python
input_network = tem.shunt_capacitor(C1) ** tem.inductor(L1)
output_network = tem.shunt_capacitor(C2) ** tem.inductor(L2) ** tem.shunt_capacitor(C3)
amplifier = input_network ** bjt ** output_network
```

The shunt keyword indicates a parallel component. Here the cascading of 2-Port networks works well. Otherwise, you can fall back to [`rf.Circuit`](https://scikit-rf.readthedocs.io/en/latest/api/circuit.html#circuit-skrf-circuit) to define more complex connection schemes.

**Reply by jhillairet** [reply]
In addition, note that you can also use the [stability circle function](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.network.Network.stability_circle.html#skrf.network.Network.stability_circle) and similar functions. We should integrate them in the example...

**Reply by stan3c** [reply]
I implemented stability circles in my script using the functions from example and it works. I did not try the stability circle function yet.
I did however try to use Network.gain_circle (bjt.gain_circle as in my post above) and that does not work as expected. I think it plots only a centre of the circle but not the circle.
I do not know what I am doing wrong. And ther is a message:
```
tmp/ipykernel_736351/3598348151.py:1: RuntimeWarning: The specified gain is greater than the maximum gain achievable by the matching network. Specify a smaller gain.
  sgc = bjt.gain_circle(target_port=1, gain=12.0
```

**Comment by stan3c** [comment]
Can someone explain why this code does not plot gain circles larger than 0.7, while  ntwk['1GHz'].max_gain returns max gain of 133?
```python
import skrf as rf
import matplotlib.pyplot as plt
ntwk = rf.Network('BFU520_05V0_010mA_NF_SP.s2p')

lsc = ntwk['1GHz'].gain_circle(target_port=1,gain=2)
rf.plotting.plot_smith(s=lsc, smith_r=1)
plt.show()

ntwk['1GHz'].max_gain
```
```
/tmp/ipykernel_525581/3284342712.py:1: RuntimeWarning: The specified gain is greater than the maximum gain achievable by the matching network. Specify a smaller gain.
  lsc = ntwk['1GHz'].gain_circle(target_port=1,gain=0.8)
```
